x/examples/soax-relay/main.go

```diff
@@ -0,0 +1,309 @@
+		errCh <- errors.New("client TCP connection closed")
+	}()
+
+	// Wait for any error or connection close
+	err = <-errCh
+	log.Printf("UDP Associate ended: %v", err)
+	return nil
+}
+
+func relayUDP(clientConn *net.UDPConn, upstreamConn *net.UDPConn, errCh chan<- error) {
+	var clientAddr *net.UDPAddr
+	var clientAddrMu sync.Mutex
+
+	// Client to upstream
+	go func() {
+		buf := make([]byte, 64*1024)
+		for {
+			n, addr, err := clientConn.ReadFromUDP(buf)
+			if err != nil {
+				errCh <- fmt.Errorf("error reading from client UDP: %v", err)
+				return
+			}
+
+			clientAddrMu.Lock()
+			if clientAddr == nil {
+				clientAddr = addr
+			}
+			clientAddrMu.Unlock()
+
+			_, err = upstreamConn.Write(buf[:n])
+			if err != nil {
+				errCh <- fmt.Errorf("error writing to upstream UDP: %v", err)
+				return
+			}
+		}
+	}()
+
+	// Upstream to client
+	go func() {
+		buf := make([]byte, 64*1024)
+		for {
+			n, err := upstreamConn.Read(buf)
+			if err != nil {
+				errCh <- fmt.Errorf("error reading from upstream UDP: %v", err)
+				return
+			}
+
+			clientAddrMu.Lock()
+			if clientAddr != nil {
+				_, err = clientConn.WriteToUDP(buf[:n], clientAddr)
+				if err != nil {
+					errCh <- fmt.Errorf("error writing to client UDP: %v", err)
+					return
+				}
+			}
+			clientAddrMu.Unlock()
+		}
+	}()
+}
+func convertToNetAddr(ip netip.Addr, port uint16) *net.UDPAddr {
+	return &net.UDPAddr{
+		IP:   ip.AsSlice(),
+		Port: int(port),
+	}
+}
+
+type Config struct {
+	Server struct {
+		TCPListenAddress string `mapstructure:"tcp_listen_address"`
+		UDPListenAddress string `mapstructure:"udp_listen_address"`
+	}
+	SOAX struct {
+		PackageID  string `mapstructure:"package_id"`
+		PackageKey string `mapstructure:"package_key"`
+		Address    string `mapstructure:"address"`
+	}
+	Credentials map[string]string `mapstructure:"credentials"`
+}
+
+func loadConfig() (*Config, error) {
+	v := viper.New()
+	v.SetConfigName("config")
+	v.SetConfigType("yaml")
+	v.AddConfigPath(".")
+
+	if err := v.ReadInConfig(); err != nil {
+		return nil, fmt.Errorf("failed to read config file: %w", err)
+	}
+
+	var config Config
+	if err := v.Unmarshal(&config); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal config: %w", err)
+	}
+
+	config.SOAX.PackageID = "package-" + config.SOAX.PackageID + "-"
+
+	if config.SOAX.Address == "" {
+		config.SOAX.Address = "proxy.soax.com:5000"
+	}
+
+	return &config, nil
+}
+
+func main() {
+
+	config, err := loadConfig()
+	if err != nil {
+		log.Fatalf("Failed to load config: %v", err)
+	}
+
+	// Create the custom authenticator
+	customAuth := CustomAuthenticator{
+		credentials: config.Credentials,
+	}
+
+	streamEndpoint := transport.StreamDialerEndpoint{Dialer: &transport.TCPDialer{}, Address: config.SOAX.Address}
+	client, err := csocks5.NewClient(&streamEndpoint)
+	if err != nil {
+		return
+	}
+
+	server := socks5.NewServer(
+		socks5.WithAuthMethods([]socks5.Authenticator{customAuth}),
```

Comment:
New Approach:
- Let's use [WithAssociateMiddleware](https://github.com/things-go/go-socks5/blob/503e7866821a419fe3766751f6d5671390072c14/option.go#L165).
  - We have access to the Request there, so we can process the credentials and create the SOCKS5 client, which we can store in the Context
  -  This is not released yet, so you will need to depend on a specific commit.
- In the Dial function (specified with `WithDial`), retrieve the client from the Context and dial.